core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -588,19 +592,20 @@ private Evaluator extractInclusiveDeleteExpression(ManifestReader reader) {
             outputManifests.add(bin.get(0));
             return;
           }
-
           // if the bin has a new manifest (the new data files) then only merge it if the number of
           // manifests is above the minimum count. this is applied only to bins with an in-memory
           // manifest so that large manifests don't prevent merging older groups.
           if (bin.contains(cachedNewManifest) && bin.size() < minManifestsCountToMerge) {
             // not enough to merge, add all manifest files to the output list
             outputManifests.addAll(bin);
+          } else if (bin.contains(cachedNewAppenedManifest) && bin.size() < minManifestsCountToMerge) {
```

Comment:
New manifests are placed at the beginning of the iterable of manifests to merge, so that newer data files are found first. That way, systems like Presto that don't wait for job planning to complete before running a query will return results more quickly. Merges preserve that order so the bin that doesn't get merged because it contains `cachedNewManifest` is the first bin, not the last.

Appended manifests are added in the order they are appended, so `add(m1).add(m2)` will result in `m1` before `m2` in the group, which could be split across bins. Using the last appended manifest for `cachedNewAppendedManifest` will cause the wrong bucket (the one with m2) to not get merged when it should be the one with m1.

You can fix this problem by tracking `firstNewManifest` instead of the last one.

Reply:
thanks a lot @rdblue for explanation !
make sense to keep the new data files or manifests in the first/top bin, and not to merge first bin if it has the latest till it reaches threshold.
 
let me change the logic to track `firstAppendedManifest` instead of `lastAppendedManifest`.

One question on this statement: 
>That way, systems like Presto that don't wait for job planning to complete before running a query will return results more quickly.

I believe we are saying that as "Presto doesn't wait for job planning to complete before running a query, so will return results **(recently added data)** more quickly,  as we are keeping the recent/new manifest files at the top" ?

Reply:
done the changes.